### ISSUE TITLE
The path to TorchScript

### ISSUE THREAD

**Issue by Luthaf** [lab-cosmo/metatensor#133]
Here is what I currently see as the path to enabling the use of equistore within TorchScript. 

For reference, TorchScript is a small, Python-like language that is implemented inside the torch library (in C++). It can be extended with new operations/classes in Python (if the new operation/class only use the TorchScript-compatible subset of Python); or in C++ (completely custom operations/classes are possible). 

To enable using equistore with TorchScript, we need to add a couple of custom classes (TensorMap, TensorBlock, Labels) to TorchScript, and we need to implement these in C++ (since we need to call into the core equistore library). This mean there will be two version of each class available from Python: the current, pure Python ones, and the TorchScript ones. See point 4 below for how to handle this. 

Once we have the basic classes and member functions (such as `keys_to_properties`) available from TorchScript, we'll have to make sure everything in `equistore.operations` is available from 

### 0) Clean separation of the core shared library

A pre-requisite to the TorchScript version of equistore (and the ability to use this version together with rascaline) is to ensure only one version of the core functions/structs exists, by splitting it into a standalone shared library that can then be loaded by both rascaline and TorchScript.

- [x] pull request: #85

### 1) Write the new TorchScript classes

We can then start writing C++ code to create new classes inheriting from `torch::CustomClassHolder`, and registering these classes with TorchScript. This means re-wrapping all classes and exposing all functions in a TorchScript-compatible way (e.g. all data uses `torch::Tensor`, …). This will first leave in a separate package (equistore_torch) in this repository. This can happen as a set of PR to a long-lived branch if we want to parallelize the work

- [X] create a long-lived branch with basic infrastructure to build and install the code #233 
- [x] implement the new classes:  #263
    - [x] Labels
    - [x] Implement [`equistore::DataArrayBase`](https://github.com/lab-cosmo/equistore/blob/3327e1b5ddd3ee9c0441e3b82a36c67044c6933f/include/equistore.hpp#L573-L582) for `torch::Tensor`
    - [x] TensorBlock/Proxy for gradients
    - [x] TensorMap
- [x] merge the long-lived branch

- [x] Implement serialization for the torch version, this requires the same core functions as #94.

### 2) Change how `Labels` works

Because all the data needs to be either a custom class or one of TorchScript type, we will have a problem with Labels. Right now, these use [numpy's structure arrays](https://numpy.org/doc/stable/user/basics.rec.html#structured-datatypes) to carry around the names of the columns in Labels and allow indexing with these columns names. Unfortunately, structured data types are not supported with torch tensors, meaning we will need to have a different API for TorchScript.

We could have a different API for standard Python & TorchScript, but I think it would be better to be able to drop in one or the other version in the same Python code, to make it easier to move between them (see point 4). So it might be better to refactor the Python version of Labels to match what the TorchScript version will be. Instead of inheriting from `np.ndarray`, the new Labels should be a standalone class, with a `values` and `names` attributes (fixing #63), with some functions to support fancy indexing & slicing of the Labels.

- [x] change the pure Python Labels class
    - [x] design the class and update the core tests
    - [x] update all operations to use the new Labels API
- [x] make sure the TorchScript and pure Python version of Labels agree

The first part of this could be done in parallel to point 1.

### 3) Ensure `equistore.operations` is usable from TorchScript

Once we have all the base classes available in TorchScript, we'll need to go through the operations one by one, and add a test that they can be exported to pure TorchScript, without any Python.

- [x] make sure the functions in `_dispatch` can be serialized to TorchScript
- [x] go over the high-level operations and port them
    - [x] `empty_like`
    - [x] `ones_like`
    - [x] `zeros_like`
    - [x] `dot`
    - [x] `lstsq`
    - [x] `solve`
    - [x] `allclose`
    - [x] `equal`
    - [x] `equal_metadata`
    - [x] `join`
    - [x] `splice`
    - [x] `split`
    - [x] `sum_over_samples`
    - [x] `mean_over_samples`
    - [x] `remove_gradients`
    - [x] `add`
    - [x] `divide`
    - [x] `multiply`
    - [x] `pow`
    - [x] `substract`
    - [x] `block_from_array`
    - [x] `drop_block`
    - [x] `one_hot`
    - [x] `to`
    - [x] `abs`
    - [x] `unique_metadata`
    - [x] `append_dimension`
    - [x] `permute_dimensions`
    - [x] `remove_dimension`
    - [x] `rename_dimension`

### 4) Select between pure Python & TorchScript versions of classes

Before or at the same time as point 3, we will also have to pick how the user should interact with the TorchScript version of the code. I can basically see two (and a half) API:

```py
# version A, "explicit"
from equistore import TensorMap  # gives the pure Python class
from equistore.torch import TensorMap  # gives the TorchScript class

# version B.1, "implicit"
from equistore import TensorMap  # if torch can be imported, gives the TorchScript class; otherwise gives the pure Python class

# version B.2, "a bit less implicit", not sure if this is possible
from equistore import TensorMap  # gives the pure Python class

# somehow tell pytorch to replace TensorMap with `equistore.torch.TensorMap` when exporting to TorchScript.
```

This is a decision we can mostly delay until step 1 is done.

**Comment by agoscinski** [comment]
Does this also allows us to use torch.nn.functional operations on TensorMaps https://pytorch.org/docs/stable/nn.functional.html ?

**Comment by Luthaf** [comment]
No, we would still have to wrap these with operations. Although for things applied on a block-by-block basis, without metadata manipulation, we could create a single operations (something like `apply(tensor, function)`).

TorchScript knowing about equistore is more that 

1. TorchScript knows there is a `TensorMap` class
2. there is a `.block().data` function that can be used on `TensorMap` and return a `torch.Tensor`
3. ... same thing for all classes and functions in equistore

This means we still have to write code to do something useful, but TorchScript will understand the metadata manipulation and data extraction parts of our code.

**Comment by Luthaf** [comment]
Regarding step 4 (`Select between pure Python & TorchScript versions of classes`), the decision is to force an explicit import like this:

```py
# using Python backend
from equistore import TensorMap
import equistore

data = equistore.load("...")
modified = equistore.dot(data, equistore.transpose(data))
# alternatively modified = data @ data.T

# using TorchScript backend
from equistore.torch import TensorMap
import equistore.torch

data = equistore.torch.load("...")
modified = equistore.torch.dot(data, equistore.torch.transpose(data))
# alternatively modified = data @ data.T
```

where `equistore.torch.dot` and `equistore.dot` are not the same function. Instead we will import equistore-operations twice in different context (defining `TensorMap` to be the one from `equistore-core` or `equistore-torch`), and re-exporting all functions from both `equistore` and `equistore.torch`. We need to define the functions twice to make sure the type `__annotations__` used by the TorchScript compiler are pointing to the correct versions of the classes.

---

This will also make it really easy to port a model to TorchScript, just changing the import

```py
import equistore
```

to 

```py
import equistore.torch as equistore
```

**Comment by agoscinski** [comment]
I played around with PR #263 and wanted to share my problems with equistore.operations so far.
- most (maybe all) np.where calls need to dispatched because Labels.values can be torch.Tensor
- tensor.Tensor.size does not exist
- unique does not support return_indices so the dispatch function must implement this, shouldn't be complicated

**Comment by Luthaf** [comment]
> We might need to remove the typings in the operations. One could also set them all to torch.TensorBlock when HAS_TORCH is True, but this would be obscure typing, since we still can do operations with numpy equistore objects

So the plan here is to have two copies of the operations, generated from the same source code. One of the copies would be in `equistore.add`/`equistore.operations.add`, and have type annotation using `equistore.core.TensorMap`. The other one would be in `equistore.torch.add`/`equistore.torch.operations.add`, and have type annotation using `equistore.torch.TensorMap`. 

This is possible by using `importlib`, which allow re-importing a module with different global context (the global context would either expose `equistore.core.TensorMap` or `equistore.torch.TensorMap`).

**Comment by Luthaf** [comment]
Now that #320 and #310 have been merged, we can start the final part of this journey, making sure all operations are compatible with torch script! There is an example of how to do this in b4969ed51d55a8a845acb4bad22ed1c86734ddc9, basically the following steps:

-  add a test file for this operation
- check that the operation still runs when called from `equistore.torch.<operation>`
- check that the operation compiles with `torch.jit.script(equistore.torch.<operation>)`, and that the resulting code runs as well. 

Please reply here to pick operations you would like to convert!

**Comment by frostedoyster** [comment]
I'll try to start with `empty_like`, `ones_like`, `zeros_like`. I can take more if that goes well

**Comment by frostedoyster** [comment]
Acutally, is it a smarter idea to start from the logic functions? These can be used to perform all sorts of checks in other tests

**Comment by agoscinski** [comment]
Hm... I don't see that operations like allclose and equal are used in other operations (only in the tests of other operations which should not be relevant for testing if it is torch scriptable). But if you find a function/operation that is not torch scriptable that is used in multiple operations, then yes please do it first. That is why we done so far the functions in `_dispatch.py `(already merged)  and `_utils.py` https://github.com/lab-cosmo/equistore/pull/322

**Comment by frostedoyster** [comment]
Actually Guillaume's sample test does test some basic correctness of the output, and these operations would be useful for that

**Comment by frostedoyster** [comment]
Ok, I'll work on it

**Comment by frostedoyster** [comment]
I'll also try `add`, `subtract`, `divide`, `multiply`, `pow` if nobody else wants them

**Comment by frostedoyster** [comment]
Claiming `reduce_over_samples` operations, if nobody else is interested

**Comment by Luthaf** [comment]
All done! 🎉
